File: python/audit.py
import abc
import os.path
from datetime import datetime
from typing import Any


class FileSystem(abc.ABC):
    @abc.abstractmethod
    def get_files(self, dir_name: str):
        raise NotImplementedError

    @abc.abstractmethod
    def write_all_text(self, new_file: str, content: str):
        raise NotImplementedError

    @abc.abstractmethod
    def read_all_lines(self, path: str):
        raise NotImplementedError


class AuditManager:
    def __init__(self, max_entries_per_file: int, directory_name: str, file_system: FileSystem):
        self._max_entries_per_file = max_entries_per_file
        self._directory_name = directory_name
        self._file_system = file_system

    def add_record(self, visitor_name: str, time_of_visit: datetime):
        file_paths = self._file_system.get_files(self._directory_name)
        sorted_paths = self._sort_by_index(file_paths)
        new_record = visitor_name + ';' + time_of_visit.strftime("%Y-%m-%d %H:%M:%S")

        if len(sorted_paths) == 0:
            new_file = os.path.join(self._directory_name, 'audit_1.txt')
            self._file_system.write_all_text(new_file, new_record)
            return

        current_file_index, curr_file_path = sorted_paths[-1]
        lines = self._file_system.read_all_lines(curr_file_path)

        if len(lines) < self._max_entries_per_file:
            lines.append(new_record)
            new_content = '\n'.join(lines)
            self._file_system.write_all_text(curr_file_path, new_content)
        else:
            new_index = current_file_index + 1
            new_name = f'audit_{new_index}.txt'
            new_file = os.path.join(self._directory_name, new_name)
            self._file_system.write_all_text(new_file, new_record)

    @staticmethod
    def _sort_by_index(file_paths) -> list[tuple[Any, Any]]:
        return list(enumerate(sorted(file_paths), start=1))

File: typescript/audit.ts
import path from "path";

export interface FileSystem {
  getFiles(directoryName: string): string[];
  writeAllText(filePath: string, content: string): void;
  readAllLines(filePath: string): string[];
}

export class AuditManager {
  private maxEntriesPerFile: number;
  private directoryName: string;
  private fileSystem: FileSystem;

  constructor(maxEntriesPerFile: number, directoryName: string, fileSystem: FileSystem) {
    this.maxEntriesPerFile = maxEntriesPerFile;
    this.directoryName = directoryName;
    this.fileSystem = fileSystem;
  }

  addRecord(visitorName: string, timeOfVisit: Date): void {
    const filePaths = this.fileSystem.getFiles(this.directoryName);
    const sortedPaths = AuditManager.sortByIndex(filePaths);
    const newRecord = `${visitorName};${timeOfVisit
      .toISOString()
      .replace("T", " ")
      .substring(0, 19)}`;

    if (sortedPaths.length === 0) {
      const newFile = path.join(this.directoryName, "audit_1.txt");
      this.fileSystem.writeAllText(newFile, newRecord);
      return;
    }

    const [currentFileIdx, currFilePath] = sortedPaths[sortedPaths.length - 1];
    const lines = this.fileSystem.readAllLines(currFilePath);

    if (lines.length < this.maxEntriesPerFile) {
      lines.push(newRecord);
      const newContent = lines.join("\n");
      this.fileSystem.writeAllText(currFilePath, newContent);
    } else {
      const newIndex = currentFileIdx + 1;
      const newName = `audit_${newIndex}.txt`;
      const newFile = path.join(this.directoryName, newName);
      this.fileSystem.writeAllText(newFile, newRecord);
    }
  }

  static sortByIndex(filePaths: string[]): [number, string][] {
    return filePaths.sort().map((path, index) => [index + 1, path]);
  }
}
